### ISSUE TITLE
Error function `erf` returns values  >1, outside of image, different to scipy

### ISSUE THREAD

**Issue by JulianGoeltz** [google/jax#14717]
### Description

The [error function](https://en.wikipedia.org/wiki/Error_function) has as its image the open interval `(-1, 1)`. Calling the `jax.scipy.special.erf` function sometimes returns values outside of this range. Even though the observed deviation is small, the values can have unfortunate repercussions because they are outside of the defined set (in my example it resulted in negative values that were supposed to be probabilities).
The original `scipy` implementation does not seem to have this issue, at least not for the same values:
```
>>> import jax, scipy
>>> x = 3.9004288
>>> print(jax.scipy.special.erf(x), scipy.special.erf(x))
No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
1.0000001 0.999999965327526
```

I tried to look at the implementation to find the cause for this issue but was not successful.

### What jax/jaxlib version are you using?

jax v0.4.4, jaxlib v0.4.4

### Which accelerator(s) are you using?

CPU

### Additional system info

Python version 3.10.9, linux 6.1.12

**Comment by jakevdp** [comment]
Thanks for the report! This looks like a floating point precision issue. If you [enable 64-bit precision](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#double-64bit-precision) then the outputs are consistent with scipy:
```python
>>> import jax, scipy
>>> jax.config.update('jax_enable_x64', True)
>>> x = 3.9004288
>>> print(jax.scipy.special.erf(x), scipy.special.erf(x))
0.999999965327526 0.999999965327526
```
The float32 results are probably a bug, but not a bad one: in general float32 has 23 bits of mantissa, so operations are good to about 1 part in 2^23 or roughly 1 part in 10^7, which is consistent with the float32 output above.

**Comment by JulianGoeltz** [comment]
Thanks for the quick reply!

Interesting, I tried the opposite (give a 32-bit array to scipy) and got a result as expected (<1). I guess I will just have to `clip` the results to the correct range.
Out of interest and if it is no effort for you, could you direct me to the point in the source code where this function is implemented?

**Comment by jakevdp** [comment]
I believe the difference is that scipy does the computation in 64-bit, then casts the output to 32-bit. In JAX's case, the operation itself is done in 32-bit.

The implementation of the 32-bit Erf operation is here: https://github.com/tensorflow/tensorflow/blob/14683e315ecbb9e831121be8b78815f6f8927a60/tensorflow/compiler/xla/client/lib/math.cc#L201-L216

**Comment by JulianGoeltz** [comment]
Thanks for taking care of this and creating the issue in the correct place! :)

**Comment by rajasekharporeddy** [comment]
Hi @JulianGoeltz 

Looks like this issue has been resolved by https://github.com/tensorflow/mlir-hlo/commit/29e975a3c0a469c41fc564c36766217aef9d07a4. I tested the mentioned code on colab CPU with JAX version 0.4.26. JAX now clamps the output of `erf` function to [-1:1].

```python
import jax, scipy
import jax.numpy as jnp
x = 3.9004288
print(jax.scipy.special.erf(x), scipy.special.erf(x))
```
Output:
```
1.0 0.999999965327526
```
Attaching the [gist](https://colab.sandbox.google.com/gist/rajasekharporeddy/09d15b40efc2a9954dde0ab91f6f8301/-14717.ipynb) for reference.

Thank you
